dbms/src/Flash/Mpp/EstablishCall.h

```diff
@@ -17,29 +17,30 @@
 #include <Common/MPMCQueue.h>
 #include <Common/Stopwatch.h>
 #include <Flash/FlashService.h>
-#include <Flash/Mpp/MPPTunnel.h>
-#include <Flash/Mpp/PacketWriter.h>
+#include <Flash/Mpp/GRPCSendQueue.h>
+#include <kvproto/tikvpb.grpc.pb.h>
 
 namespace DB
 {
-class MPPTunnel;
 class AsyncTunnelSender;
 class AsyncFlashService;
 
-class SyncPacketWriter : public PacketWriter
+class IAsyncCallData
 {
 public:
-    explicit SyncPacketWriter(grpc::ServerWriter<mpp::MPPDataPacket> * writer)
-        : writer(writer)
-    {}
+    virtual ~IAsyncCallData() = default;
 
-    bool write(const mpp::MPPDataPacket & packet) override { return writer->Write(packet); }
+    /// Should return a non-null `grpc_call`.
+    virtual grpc_call * grpcCall() = 0;
 
-private:
-    ::grpc::ServerWriter<::mpp::MPPDataPacket> * writer;
+    /// Attach async sender in order to notify consumer finish msg directly.
+    virtual void attachAsyncTunnelSender(const std::shared_ptr<DB::AsyncTunnelSender> &) = 0;
```

Comment:
`AsyncTunnelSender` must outlive `IAsyncCallData`.

Reply:
Then I thought a raw pointer is enough? Or `IAsyncCallData` may mistakenly extend the lifetime of `AsyncTunnelSender`.

Reply:
My last comment is not quite correct, but here `shared_ptr` make the relationship too complicated (at least from the first glance).

Reply:
I am wrong. In the current implementation, the `consumerFinish` is called before calling `responder.Finish`. Then `EstablishCallData` will wait for the event for `responder.Finish` then it will destroy itself. For now, `AsyncTunnelSender` does not touched in `Finish` state.
I don't know if calling `consumerFinish` after `responder.Finish` is better than before.